docs/design/2022-06-15-phyical_plan_explain.md

```diff
@@ -0,0 +1,472 @@
+        isIndexScan=true;
+        break;
+      case COVERING_INDEX_SCAN:
+        sb.append("CoveringIndexScan");
+        break;
+      case TABLE_SCAN:
+        sb.append("TableScan");
+    }
+    if (isIndexScan && !getDowngradeFilters().isEmpty()) {
+       sb.append(", Downgrade Filter: ");
+       Joiner.on(", ").skipNulls().appendTo(sb, request.getDowngradeFilters());
+    }
+    if (!isIndexScan && !getFilters().isEmpty()) {
+        sb.append(", Residual Filter: ");
+        Joiner.on(", ").skipNulls().appendTo(sb, request.getFilters());
+    }
+    if (!getPushDownFilters().isEmpty()) {
+        sb.append(", PushDown Filter: ");
+        Joiner.on(", ").skipNulls().appendTo(sb, request.getPushDownFilters());
+    }
+    // add keyRange and other push down expression to sb.
+    ...
+    return sb.toString();
+}
+```
+
+From the code we can see that the execution plan is divided into three kinds of `IndexScan`, `CoveringIndexScan`, and `TableScan`; `TiDAGRequest` has three kinds `Filter` —— `Downgrade Filter`, ` Resdiual Filter`, `PushDown Filter`. There is zero or more `Filter` for an execution plan. We will describe the meaning of each execution plan and `Filter` in detail below.
+
+### Meaning of execution plan
+
+#### `IndexScan`
+
+In TiSpark, an `IndexScan` requires two scans. The first scan is scanning in index data and the second scan is scanning in table data to get the data we need from the results returned in the first scan.
+
+#### `CoveringIndexScan`
+
+`CoveringIndexScan` is a special case of `IndexScan`. If the column(s) in `Filter` and the column(s) in `Projection` are both inside the Index, then we only need to scan the index once, no need to scan for the table. That is, it only scans for the index, such a scan is called `CoveringIndexScan`.
+
+#### `TableScan`
+
+`TableScan` is different from `IndexScan` and `CoveringIndexScan`. `TableScan` only scans table data.
+
+### Meaning of `Filter`
+
+#### `Downgrade Filter`
+
+If the second scan in `IndexScan` does too many queries on COP([TiKV Coprocessor](https://docs.pingcap.com/tidb/stable/tikv-overview#tikv-coprocessor))/TiKV, it can cause performance problems in COP/TiKV. To solve this problem a downgrading mechanism is introduced. `Downgrade Filter` is used after `IndexScan` downgrade.
+
+The first scan of `IndexScan` will return the data that meets the conditions, and then TiSpark will sort and aggregate the returned data to obtain the `regionTask` that needs to be done in the `IndexScan` second scan. If the number of `regionTask` is bigger than `downgradeThreshold`, a downgrade will be triggered. When a downgrade is triggered, the range of the second scan table will be changed from the returned value of the first scan index to all values in the middle of the minimum to maximum value of the first scan index, and the `filters` of the second scan will change to `downgradeFilters`(`downgradeFilters` is the same as if the execution plan is `TableScan`'s `filters`).
```

Comment:
How about changing "in the middle of the minimum to maximum value of the first scan index" to "between the minimum and maximum value of the first scan index "?

Reply:
changed